pkg/p2consul/kv.go

```diff
@@ -0,0 +1,138 @@
+// Package p2consul provides a generalized API for reading and writing manifests
+// in consul.
+package p2consul
+
+import (
+	"bytes"
+	"strings"
+	"time"
+
+	"github.com/armon/consul-api"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/util"
+)
+
+type ManifestResult struct {
+	Manifest *pods.PodManifest
+	Path     string
+}
+type Options struct {
+	Token   string
+	Address string
+}
+
+type Store struct {
+	client *consulapi.Client
+}
+
+// Slash joins its arguments with slashes and returns the result.
+func Slash(args ...string) string {
+	return strings.Join(args, "/")
+}
+
+func NewStore(opts Options) *Store {
+	conf := consulapi.DefaultConfig()
+	conf.Address = opts.Address
+	conf.Token = opts.Token
+
+	// the error is always nil
+	client, _ := consulapi.NewClient(conf)
+	return &Store{client: client}
+}
+
+// SetPod writes a pod manifest into the consul key-value store. The key should
+// not have a leading or trailing slash.
+func (s *Store) SetPod(key string, manifest pods.PodManifest) (time.Duration, error) {
+	buf := bytes.Buffer{}
+	err := manifest.Write(&buf)
+	if err != nil {
+		return 0, err
+	}
+	keyPair := &consulapi.KVPair{
+		Key:   key,
+		Value: buf.Bytes(),
+	}
+
+	writeMeta, err := s.client.KV().Put(keyPair, nil)
+	if writeMeta == nil {
```

Comment:
the error gets returned anyway, so I don't think it matters. The bigger issue there is that, if there was an error, writemeta might be nil so derefing it causes a panic.